README.md
# mpesa-sdk

Dart package for VODACOM M-Pesa API(Open API) (TANZANIA,GHANA AND LESOTHO )
MARKET = vodacomTZN, vodafoneGHA, vodacomLES
COUNTRY = TZN, GHA, LES
CURRENCY = TZS, GHS, LSL


## Features
Ready Methods/APIs

- [√] C2B
- [√] B2C
- [√] B2B
- [√] REVERSAL
- [√] TRANSACTION STATUS



## Getting started

### Credentials

1. Create an account on the [Vodacom Developer Portal](https://openapiportal.m-pesa.com/)
2. Register Your application
3. Get your keys -> `apiKey` and `publicKey`

## Usage
Add dependency on pubspec.yaml
```yaml
dependencies:
  mpesa_sdk: <latest_version>
```
Import in your Flutter app or plain dart app.

```dart
import 'package:mpesa_sdk/mpesa_sdk.dart';
```
Create an instance of Mpesa, then use its methods to consume the Api

```dart
import 'package:mpesa_sdk/mpesa_sdk.dart';
void main() async{
  String publicKey = "";
  String apiKey= "";

  final mpesa = Mpesa(
      env: ENV.sandbox,
      market: MarketCountry.vodacomTZN,
      apiKey: apiKey,
      publicKey: publicKey
  );

  final res = await mpesa.c2b(
      inputAmount: 1000,
      inputCustomerMSISDN: "000000000001", // CUSTOMER PHONE NUMBER
      inputServiceProviderCode: "000000", // PROVIDED CODE BY VODACOM AFTER REGISTERING YOUR ORG.
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID: "asv02e5yy58774fttt228d83d0d689761", // UNIQUE
      inputTransactionReference: "T12774C", // UNIQUE
      inputPurchasedItemsDesc: "BAG");

  print("********C2B**********<<<<<<<<<<<${res?.getBody()}>>>>>>>>>>>**********************C2B********");


  final res2 = await mpesa.b2c(
      inputAmount: 1000,
      inputCustomerMSISDN: "000000000001",
      inputServiceProviderCode: "000000",
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID: "as2v02e5yy58774fttt228d83d0d6897613444",
      inputTransactionReference: "T127744",
      inputPaymentItemsDesc: "SALARY PAYMENT");

  print("********B2C**********<<<<<<<<<<<${res2?.getBody()}>>>>>>>>>>>**********************B2C********");


  final res3 = await mpesa.b2b(
      inputAmount: 1000,
      inputReceiverPartyCode: "000001",
      inputPrimaryPartyCode: "000000",
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID: "as3v02e5yy58774fttt228d83d0d68976155",
      inputTransactionReference: "T1267755",
      inputPaymentItemsDesc: "SALARY PAYMENT");

  print("********B2B**********<<<<<<<<<<<${res3?.getBody()}>>>>>>>>>>>**********************B2B********");

  final res4 = await mpesa.reversal(
      inputReversalAmount: 25,
      inputServiceProviderCode: "000000",
      inputThirdPartyConversationID: "asv02e5958774f7ba228d83d0d689777",
      inputTransactionID: "0000000000001",
      inputCountry: "TZN"
  );

  print("********REVERSAL**********<<<<<<<<<<<${res4?.getBody()}>>>>>>>>>>>**********************REVERSAL********");

  final res5 = await mpesa.getTransactionStatus(
      inputQueryReference: "000000000000000000001",
      inputServiceProviderCode: "000000",
      inputThirdPartyConversationID: "asv02e5958774f7ba228d83d0d689761",
      inputCountry: "TZN"
  );


  print(
      "********Transaction Status**********<<<<<<<<<<<${res5?.getBody()}>>>>>>>>>>>**********************Transaction Status********");


}
```

## Copyright and license

MIT License

Copyright (c) 2022 Sadick F Mghase

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
## 0.0.1

*  initial release, only c2b supported

## 0.0.1+1
* Added b2c, b2b and transaction reversal

## 0.1.0
* Added transaction status
* Fix minor issues
* Added comment documentations in api's

test/mpesa_sdk_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mpesa_sdk/mpesa_sdk.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/src/mpesa.dart
import 'package:dio/dio.dart';
import 'package:mpesa_sdk/mpesa_sdk.dart';

/// environment mode for testing [ENV.sandbox] and for production [ENV.openapi]
enum ENV { sandbox, openapi }

/// Currently supported country. Tanzania [MarketCountry.vodacomTZN], GHANA [MarketCountry.vodafoneGHA]
/// and LESOTHO [MarketCountry.vodacomLES]
enum MarketCountry { vodacomTZN, vodafoneGHA, vodacomLES }

class Mpesa {
  final String apiKey;
  final String publicKey;
  final ENV env;
  final MarketCountry market;
  late APIContext apiContext;
  late APIRequest req;
  APIResponse? response;

  /// Constructor with optional parameters
  Mpesa(
      {this.env = ENV.sandbox,
      this.market = MarketCountry.vodacomTZN,
      required this.apiKey,
      required this.publicKey});

  /// Customer to business
  Future<APIResponse?> c2b(
      {required double inputAmount,
      required String inputCustomerMSISDN,
      String inputCountry = "TZN",
      String inputCurrency = "TZS",
      required String inputServiceProviderCode,
      required String inputThirdPartyConversationID,
      required String inputTransactionReference,
      required String inputPurchasedItemsDesc,
      String? sessionId}) async {
    final sessionID = await getSession(sessionId: sessionId);
    apiContext = APIContext(headers: {}, parameters: {})
      ..setApiKey(sessionID!)
      ..setPublicKey(publicKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.POST)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/c2bPayment/singleStage/");

    apiContext.addParameter("input_Amount", inputAmount);
    apiContext.addParameter("input_Country", inputCountry);
    apiContext.addParameter("input_Currency", inputCurrency);
    apiContext.addParameter("input_CustomerMSISDN", inputCustomerMSISDN);
    apiContext.addParameter(
        "input_ServiceProviderCode", inputServiceProviderCode);
    apiContext.addParameter(
        "input_ThirdPartyConversationID", inputThirdPartyConversationID);
    apiContext.addParameter(
        "input_TransactionReference", inputTransactionReference);
    apiContext.addParameter(
        "input_PurchasedItemsDesc", inputPurchasedItemsDesc);
    req = APIRequest(apiContext);
    apiContext.addHeader("Origin", "*");
    await Future.delayed(const Duration(seconds: 1));
    try {
      response = await req.execute();
      return response;
    } on DioError catch (e) {
      throw Exception(e.response?.data);
    }
  }

  /// Business to customer
  Future<APIResponse?> b2c(
      {required double inputAmount,
      required String inputCustomerMSISDN,
      String inputCountry = "TZN",
      String inputCurrency = "TZS",
      required String inputServiceProviderCode,
      required String inputThirdPartyConversationID,
      required String inputTransactionReference,
      required String inputPaymentItemsDesc,
      String? sessionId}) async {
    final sessionID = await getSession(sessionId: sessionId);
    apiContext = APIContext(headers: {}, parameters: {})
      ..setApiKey(sessionID!)
      ..setPublicKey(publicKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.POST)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/b2cPayment/");

    apiContext.addParameter("input_Amount", inputAmount);
    apiContext.addParameter("input_Country", inputCountry);
    apiContext.addParameter("input_Currency", inputCurrency);
    apiContext.addParameter("input_CustomerMSISDN", inputCustomerMSISDN);
    apiContext.addParameter(
        "input_ServiceProviderCode", inputServiceProviderCode);
    apiContext.addParameter(
        "input_ThirdPartyConversationID", inputThirdPartyConversationID);
    apiContext.addParameter(
        "input_TransactionReference", inputTransactionReference);
    apiContext.addParameter("input_PaymentItemsDesc", inputPaymentItemsDesc);
    apiContext.addHeader("Origin", "*");
    req = APIRequest(apiContext);

    /// SessionID can take up to 30 seconds to become 'live' in the system and will be invalid until it is
    await Future.delayed(const Duration(seconds: 1)); //30
    try {
      response = await req.execute();
      return response;
    } on DioError catch (e) {
      throw Exception(e.message);
    }
  }

  /// Business to business
  Future<APIResponse?> b2b(
      {required double inputAmount,
      required String inputReceiverPartyCode,
      required String inputPrimaryPartyCode,
      required String inputThirdPartyConversationID,
      required String inputTransactionReference,
      required String inputPaymentItemsDesc,
      String inputCountry = "TZN",
      String inputCurrency = "TZS",
      String? sessionId}) async {
    final sessionID = await getSession(sessionId: sessionId);
    apiContext = APIContext(headers: {}, parameters: {})
      ..setApiKey(sessionID!)
      ..setPublicKey(publicKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.POST)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/b2bPayment/");

    apiContext.addParameter("input_Amount", inputAmount);
    apiContext.addParameter("input_Country", inputCountry);
    apiContext.addParameter("input_Currency", inputCurrency);
    apiContext.addParameter("input_ReceiverPartyCode", inputReceiverPartyCode);
    apiContext.addParameter("input_PrimaryPartyCode", inputPrimaryPartyCode);
    apiContext.addParameter(
        "input_ThirdPartyConversationID", inputThirdPartyConversationID);
    apiContext.addParameter(
        "input_TransactionReference", inputTransactionReference);
    apiContext.addParameter("input_PurchasedItemsDesc", inputPaymentItemsDesc);
    apiContext.addHeader("Origin", "*");
    req = APIRequest(apiContext);

    /// SessionID can take up to 30 seconds to become 'live' in the system and will be invalid until it is
    await Future.delayed(const Duration(seconds: 1)); //30
    try {
      response = await req.execute();
      return response;
    } on DioError catch (e) {
      throw Exception(e.message);
    }
  }

  /// Transaction reversal
  Future<APIResponse?> reversal(
      {required double inputReversalAmount,
      required String inputServiceProviderCode,
      required String inputThirdPartyConversationID,
      required String inputTransactionID,
      String inputCountry = "TZN",
      String? sessionId}) async {
    final sessionID = await getSession(sessionId: sessionId);
    apiContext = APIContext(headers: {}, parameters: {})
      ..setApiKey(sessionID!)
      ..setPublicKey(publicKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.PUT)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/reversal/");

    apiContext.addParameter("input_ReversalAmount", inputReversalAmount);
    apiContext.addParameter("input_Country", inputCountry);
    apiContext.addParameter(
        "input_ServiceProviderCode", inputServiceProviderCode);
    apiContext.addParameter(
        "input_ThirdPartyConversationID", inputThirdPartyConversationID);
    apiContext.addParameter("input_TransactionID", inputTransactionID);

    apiContext.addHeader("Origin", "*");
    req = APIRequest(apiContext);

    /// SessionID can take up to 30 seconds to become 'live' in the system and will be invalid until it is
    await Future.delayed(const Duration(seconds: 1)); //30
    try {
      response = await req.execute();
      return response;
    } on DioError catch (e) {
      throw Exception(e.message);
    }
  }

  /// Get transaction status
  Future<APIResponse?> getTransactionStatus(
      {required String inputQueryReference,
      required String inputServiceProviderCode,
      required String inputThirdPartyConversationID,
      String inputCountry = "TZN",
      String? sessionId}) async {
    final sessionID = await getSession(sessionId: sessionId);
    apiContext = APIContext(headers: {}, parameters: {})
      ..setApiKey(sessionID!)
      ..setPublicKey(publicKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.GET)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/queryTransactionStatus/");

    apiContext.addParameter("input_QueryReference", inputQueryReference);
    apiContext.addParameter(
        "input_ServiceProviderCode", inputServiceProviderCode);
    apiContext.addParameter(
        "input_ThirdPartyConversationID", inputThirdPartyConversationID);
    apiContext.addParameter("input_Country", inputCountry);

    apiContext.addHeader("Origin", "*");
    req = APIRequest(apiContext);

    /// SessionID can take up to 30 seconds to become 'live' in the system and will be invalid until it is
    await Future.delayed(const Duration(seconds: 1)); //30
    try {
      response = await req.execute();
      return response;
    } on DioError catch (e) {
      throw Exception(e.message);
    }
  }

  /// Get session
  Future<String?> getSession({String? sessionId}) async {
    if (sessionId != null) return sessionId;

    apiContext = APIContext(headers: {}, parameters: {})
      ..setPublicKey(publicKey)
      ..setApiKey(apiKey)
      ..setSSL(true)
      ..setMethodType(APIMethodType.GET)
      ..setAddress("openapi.m-pesa.com")
      ..setPort(443)
      ..setPath("/${env.name}/ipg/v2/${market.name}/getSession/");
    apiContext.addHeader("Origin", "*");

    req = APIRequest(apiContext);
    // apiContext.addHeader("Origin", "*");
    // Do the API call and put result in a response packet
    //  APIResponse? response;
    try {
      response = await req.execute();
      sessionId = response?.getBody()["output_SessionID"];
      return sessionId;
    } on DioError catch (e) {
      throw Exception(e.message);
    }
  }
}


lib/src/crypt_sra.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:basic_utils/basic_utils.dart';
import 'package:pointycastle/export.dart';

class CryptSRA {
  ///function to encrypt [apiKey] and [publicKey]
  static String encrypt({required String apiKey, required String publicKey}) {
    var pem =
        '-----BEGIN RSA PUBLIC KEY-----\n$publicKey\n-----END RSA PUBLIC KEY-----';
    var public = CryptoUtils.rsaPublicKeyFromPem(pem);

    /// Initializing Cipher
    var cipher = PKCS1Encoding(RSAEngine())
      ..init(true, PublicKeyParameter<RSAPublicKey>(public));

    /// Converting into a [Unit8List] from List<int>
    /// Then Encoding into Base64
    Uint8List output = cipher.process(Uint8List.fromList(utf8.encode(apiKey)));
    var base64EncodedText = base64Encode(output);
    return base64EncodedText;
  }
}


lib/src/api_context.dart
import 'api_method_type.dart';

/// API Context that contain info for the API endpoint
class APIContext {
  String apiKey;
  String publicKey;
  bool ssl;
  int methodType;
  String address;
  int port;
  String path;
  Map<String, dynamic> headers;
  Map<String, dynamic> parameters;

  /// Constructor with optional pre-populated variables
  APIContext({
    this.apiKey = '',
    this.publicKey = '',
    this.ssl = false,
    this.methodType = APIMethodType.GET,
    this.address = '',
    this.port = 80,
    this.path = '',
    required this.headers,
    required this.parameters,
  }); // = [];

  /// Build the URL from context parameters
  String get getUrl {
    if (ssl == true) {
      return 'https://$address:$port$path';
    } else {
      return 'http://$address:$port$path';
    }
  }

  /// Add/update headers
  void addHeader(String header, dynamic value) {
    headers[header] = value;
  }

  /// Get headers as an array
  Map<String, dynamic> getHeaders() {
    return headers;
  }

  /// Add parameter
  void addParameter(String key, dynamic value) {
    parameters[key] = value;
  }

  /// Get parameters
  Map<String, dynamic> getParameters() {
    return parameters;
  }

  /// get api key
  String getApiKey() {
    return apiKey;
  }

  /// set api key
  void setApiKey(String apiKey) {
    if (apiKey.runtimeType != String) {
      throw Exception('api_key must be a string');
    } else {
      this.apiKey = apiKey;
    }
  }

  /// get public key
  String getPublicKey() {
    return publicKey;
  }

  /// set public key
  void setPublicKey(String publicKey) {
    if (publicKey.runtimeType != String) {
      throw Exception('public_key must be a string');
    } else {
      this.publicKey = publicKey;
    }
  }

  /// get ssl
  bool getSSL() {
    return ssl;
  }

  /// set ssl
  void setSSL(bool ssl) {
    if (ssl.runtimeType != bool) {
      throw Exception('ssl must be a boolean');
    } else {
      this.ssl = ssl;
    }
  }

  /// Get method type
  int getMethodType() {
    return methodType;
  }

  /// Set methode type [0][1][2] for GET POST AND PUT
  void setMethodType(int methodType) {
    if (methodType.runtimeType != int) {
      throw Exception('method_type must be a integer');
    } else {
      this.methodType = methodType;
    }
  }

  /// get address
  String getAddress() {
    return address;
  }

  ///set address
  void setAddress(String address) {
    if (address.runtimeType != String) {
      throw Exception('address must be a string');
    } else {
      this.address = address;
    }
  }

  /// Get port number
  int getPort() {
    return port;
  }

  ///  Set port number
  void setPort(int port) {
    if (port.runtimeType != int) {
      throw Exception('port must be a integer');
    } else {
      this.port = port;
    }
  }

  /// get path endpoint
  String getPath() {
    return path;
  }

  ///set path
  void setPath(String path) {
    if (path.runtimeType != String) {
      throw Exception('path must be a string');
    } else {
      this.path = path;
    }
  }
}


lib/src/api_request.dart
import 'package:mpesa_sdk/src/api_method_type.dart';
import 'package:mpesa_sdk/src/api_response.dart';
import 'package:mpesa_sdk/src/crypt_sra.dart';
import 'package:dio/dio.dart';

/// Creates the API Request from the context
class APIRequest {
  var context;

  /// Constructor context
  APIRequest(this.context);

  /// Does the HTTP Request
  Future<APIResponse?> execute() async {
    if (context == null) {
      throw Exception('Context cannot be null');
    }
    createDefaultHeaders();
    switch (context.getMethodType()) {
      case APIMethodType.GET:
        return await __get();
      case APIMethodType.POST:
        return await __post();
      case APIMethodType.PUT:
        return await __put();
      default:
        return null;
    }
  }

  /// Creates the Authorisation bearer token using the RSA public key provided
  String createBearerToken() {
    final encryptedMSG =
        CryptSRA.encrypt(apiKey: context.apiKey, publicKey: context.publicKey);
    return encryptedMSG;
  }

  /// Add the default headers
  void createDefaultHeaders() {
    context.addHeader('Authorization', 'Bearer ${createBearerToken()}');
    context.addHeader('Content-Type', 'application/json');
    context.addHeader('Accept', 'application/json');
    context.addHeader('Host', context.getAddress());
  }

  /// Do a GET request
  Future<APIResponse> __get() async {
    final dio = Dio();

    final r = await dio.get(
      context.getUrl,
      queryParameters: context.getParameters(),
      options: Options(
        headers: context.getHeaders(),
      ),
    );

    return APIResponse(
        statusCode: r.statusCode, body: r.data, headers: r.headers.map);
  }

  /// Do a POST request
  Future<APIResponse> __post() async {
    final dio = Dio();
    final r = await dio.post(
      context.getUrl,
      data: context.getParameters(),
      options: Options(
        headers: context.getHeaders(),
      ),
    );

    return APIResponse(
        statusCode: r.statusCode, body: r.data, headers: r.headers.map);
  }

  /// Do a PUT request
  Future<APIResponse> __put() async {
    final dio = Dio();

    final r = await dio.put(
      context.getUrl,
      data: context.getParameters(),
      options: Options(
        headers: context.getHeaders(),
      ),
    );

    return APIResponse(
        statusCode: r.statusCode, body: r.data, headers: r.headers.map);
  }
}


lib/src/api_response.dart
/// API Response
class APIResponse {
  int? statusCode;
  Map<String, dynamic> headers;
  Map<String, dynamic> body;

  /// Constructor
  APIResponse({this.statusCode, required this.headers, required this.body});

  ///Get header status code
  int? getStatusCode() {
    return statusCode;
  }

  /// set status code
  void setStatusCode(int statusCode) {
    if (statusCode.runtimeType != int) {
      throw Exception('status_code must be a integer');
    } else {
      this.statusCode = statusCode;
    }
  }

  /// Get herders
  Map<String, dynamic> getHeaders() {
    return headers;
  }

  /// Add headers
  void setHeaders(Map<String, dynamic> headers) {
    if (headers.runtimeType != Map) {
      throw Exception('headers must be a map');
    } else {
      this.headers = headers;
    }
  }

  /// Get response body
  Map<String, dynamic> getBody() {
    return body;
  }

  /// Set response body
  void setBody(Map<String, dynamic> body) {
    if (body.runtimeType != Map) {
      throw Exception('body must be a map');
    } else {
      this.body = body;
    }
  }
}


example/lib/main.dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:mpesa_sdk/mpesa_sdk.dart';

void main() async {
  String publicKey =
      "MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEArv9yxA69XQKBo24BaF/D+fvlqmGdYjqLQ5WtNBb5tquqGvAvG3WMFETVUSow/LizQalxj2ElMVrUmzu5mGGkxK08bWEXF7a1DEvtVJs6nppIlFJc2SnrU14AOrIrB28ogm58JjAl5BOQawOXD5dfSk7MaAA82pVHoIqEu0FxA8BOKU+RGTihRU+ptw1j4bsAJYiPbSX6i71gfPvwHPYamM0bfI4CmlsUUR3KvCG24rB6FNPcRBhM3jDuv8ae2kC33w9hEq8qNB55uw51vK7hyXoAa+U7IqP1y6nBdlN25gkxEA8yrsl1678cspeXr+3ciRyqoRgj9RD/ONbJhhxFvt1cLBh+qwK2eqISfBb06eRnNeC71oBokDm3zyCnkOtMDGl7IvnMfZfEPFCfg5QgJVk1msPpRvQxmEsrX9MQRyFVzgy2CWNIb7c+jPapyrNwoUbANlN8adU1m6yOuoX7F49x+OjiG2se0EJ6nafeKUXw/+hiJZvELUYgzKUtMAZVTNZfT8jjb58j8GVtuS+6TM2AutbejaCV84ZK58E2CRJqhmjQibEUO6KPdD7oTlEkFy52Y1uOOBXgYpqMzufNPmfdqqqSM4dU70PO8ogyKGiLAIxCetMjjm6FCMEA3Kc8K0Ig7/XtFm9By6VxTJK1Mg36TlHaZKP6VzVLXMtesJECAwEAAQ==";
  String apiKey = "eRJuTmYzqttBt4V39fEQSdnZrdLcU0N8";

  final mpesa = Mpesa(
      env: ENV.sandbox,
      market: MarketCountry.vodacomTZN,
      apiKey: apiKey,
      publicKey: publicKey);

  final res = await mpesa.c2b(
      inputAmount: 1000,
      inputCustomerMSISDN: "000000000001", // CUSTOMER PHONE NUMBER
      inputServiceProviderCode:
          "000000", // PROVIDED CODE BY VODACOM AFTER REGISTERING YOUR ORG.
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID:
          "asv02e5yy58774Z228d83d0d689761", // UNIQUE
      inputTransactionReference: "T1Z774C", // UNIQUE
      inputPurchasedItemsDesc: "BAG");

  print(
      "********C2B**********<<<<<<<<<<<${res?.getBody()}>>>>>>>>>>>**********************C2B********");

  final res2 = await mpesa.b2c(
      inputAmount: 1000,
      inputCustomerMSISDN: "000000000001",
      inputServiceProviderCode: "000000",
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID: "as2v02e5yy58774P228d83d0d6897613444",
      inputTransactionReference: "T1P7744",
      inputPaymentItemsDesc: "SALARY PAYMENT");

  print(
      "********B2C**********<<<<<<<<<<<${res2?.getBody()}>>>>>>>>>>>**********************B2C********");

  final res3 = await mpesa.b2b(
      inputAmount: 1000,
      inputReceiverPartyCode: "000001",
      inputPrimaryPartyCode: "000000",
      inputCountry: "TZN",
      inputCurrency: "TZS",
      inputThirdPartyConversationID: "as3v02e5yy587749028d83d0d68976155",
      inputTransactionReference: "T1297755",
      inputPaymentItemsDesc: "SALARY PAYMENT");

  print(
      "********B2B**********<<<<<<<<<<<${res3?.getBody()}>>>>>>>>>>>**********************B2B********");

  final res4 = await mpesa.reversal(
      inputReversalAmount: 25,
      inputServiceProviderCode: "000000",
      inputThirdPartyConversationID: "asv02e595804f7ba228d83d0d689777",
      inputTransactionID: "0000000000001",
      inputCountry: "TZN");

  print(
      "********REVERSAL**********<<<<<<<<<<<${res4?.getBody()}>>>>>>>>>>>**********************REVERSAL********");

  final res5 = await mpesa.getTransactionStatus(
      inputQueryReference: "000000000000000000001",
      inputServiceProviderCode: "000000",
      inputThirdPartyConversationID: "asv02e5958774f7ba228d83d0d689761",
       inputCountry: "TZN"
      );


  print(
      "********Transaction Status**********<<<<<<<<<<<${res5?.getBody()}>>>>>>>>>>>**********************Transaction Status********");



  //
  //  // Cryptom crypto = Cryptom(publicKey);
  //  APIContext apiContext = APIContext(headers: {}, parameters: {})
  //    ..setPublicKey(publicKey)
  //    ..setApiKey(apiKey)
  //    ..setSSL(true)
  //    ..setMethodType(APIMethodType.GET)
  //    ..setAddress("openapi.m-pesa.com")
  //    ..setPort(443)
  //    ..setPath("/sandbox/ipg/v2/vodacomTZN/getSession/");
  //  apiContext.addHeader("Origin", "*");
  //
  //
  //  APIRequest req = APIRequest(apiContext);
  //  // apiContext.addHeader("Origin", "*");
  //  // Do the API call and put result in a response packet
  //  APIResponse? response;
  //  try {
  //    response = await req.execute();
  //    print("${response?.body}");
  //
  //    print("*******************************");
  //  } on DioError catch (e) {
  //    print("Call failed: ${e.response}");
  //  }
  //
  //  String sessionId = response?.getBody()["output_SessionID"];
  //  print("SESSION_ID==${sessionId}");
  //
  //  apiContext = APIContext(headers: {}, parameters: {})
  //    ..setApiKey(sessionId)
  //    ..setPublicKey(publicKey)
  //    ..setSSL(true)
  //    ..setMethodType(APIMethodType.POST)
  //    ..setAddress("openapi.m-pesa.com")
  //    ..setPort(443)
  //    ..setPath("/sandbox/ipg/v2/vodacomTZN/c2bPayment/singleStage/");
  //
  //  apiContext.addParameter("input_Amount", "1000");
  //  apiContext.addParameter("input_Country", "TZN");
  //  apiContext.addParameter("input_Currency", "TZS");
  //  apiContext.addParameter("input_CustomerMSISDN", "000000000001");
  //  apiContext.addParameter("input_ServiceProviderCode", "000000");
  //  apiContext.addParameter("input_ThirdPartyConversationID", "asv02eYT6761");
  // apiContext.addParameter("input_TransactionReference", "T331");
  //  apiContext.addParameter("input_PurchasedItemsDesc", "Shoes GH");
  //
  //  print("-params-${apiContext.getUrl}");
  //
  //  req = APIRequest(apiContext);
  //  apiContext.addHeader("Origin", "*");
  //  await Future.delayed(const Duration(seconds: 1));
  //  try {
  //    response = await req.execute();
  //    print("--${response?.body}");
  //    print("--${response?.statusCode}");
  //    print("--${response?.headers}");
  //  } on DioError catch (e) {
  //    print("Call 2 failed: ${e.response}");
  //  }

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Scaffold(body: Text('Flutter Demo Home Page')),
    );
  }
}
